src/hooks.rs

```diff
@@ -0,0 +1,396 @@
+
+    fn hooks() -> impl Iterator<Item = Hook<CommandKind>> {
+        Self::iter()
+            .flat_map(|event| {
+                event
+                    .supported_when()
+                    .into_iter()
+                    .map(move |when| (event, when))
+            })
+            .flat_map(|(event, when)| {
+                event
+                    .supported_kind(when)
+                    .into_iter()
+                    .map(move |kind| Hook {
+                        event,
+                        when,
+                        command: kind,
+                    })
+            })
+    }
+}
+
+/// When to run a hook in relation to a given [`LifecycleEvent`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, strum::Display, strum::AsRefStr)]
+#[strum(serialize_all = "kebab-case")]
+pub enum When {
+    /// Run the hook before the event.
+    Before,
+    /// Run the hook at the event.
+    ///
+    /// This is used for the test hook.
+    During,
+    /// Run the hook after the event.
+    After,
+}
+
+/// The kind of hook; a `ghci` command or a shell command.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, strum::Display, strum::AsRefStr)]
+#[strum(serialize_all = "kebab-case")]
+pub enum CommandKind {
+    /// A shell command.
+    Shell,
+    /// A `ghci` command.
+    ///
+    /// Can either be Haskell code to run (`TestMain.test`) or a `ghci` command (`:set args ...`).
+    Ghci,
+}
+
+impl CommandKind {
+    fn placeholder_name(&self) -> &'static str {
+        match self {
+            CommandKind::Ghci => "GHCI_CMD",
+            CommandKind::Shell => "SHELL_CMD",
+        }
+    }
+}
+
+/// A command to run for a hook.
+#[derive(Debug, Clone)]
+pub enum Command {
+    /// A shell command.
+    Shell(MaybeAsyncCommand),
+    /// A `ghci` command.
+    Ghci(GhciCommand),
+}
+
+impl Command {
+    fn kind(&self) -> CommandKind {
+        match self {
+            Command::Ghci(_) => CommandKind::Ghci,
+            Command::Shell(_) => CommandKind::Shell,
+        }
+    }
+}
+
+impl Display for Command {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Command::Ghci(command) => command.fmt(f),
+            Command::Shell(command) => command.fmt(f),
+        }
+    }
+}
+
+/// A lifecycle hook, specifying a command to run and an event to run it at.
+#[derive(Debug, Clone)]
+pub struct Hook<C> {
+    /// The event to run this hook on.
+    pub event: LifecycleEvent,
+    /// When to run the hook in relation to the event.
+    pub when: When,
+    /// The command to run.
+    pub command: C,
+}
+
+impl<C> Display for Hook<C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.when {
+            When::During => write!(f, "{}", self.event.as_ref()),
+            _ => write!(f, "{}-{}", self.when, self.event),
+        }
+    }
+}
+
+impl<C> Hook<C> {
+    fn with_command<C2>(&self, command: C2) -> Hook<C2> {
```

Comment:
Is the `C2` necessary? `C` isn't constrained in any way, so I think it would always unify with whatever the type of `command` is.

Reply:
Yeah, because this lets you change the generic type (i.e. it lets you go from a `Hook<CommandKind>` to a `Hook<Command>`). That's also why I can't use the `..self` record update syntax here — the types need to match exactly.